### PR TITLE
[wip] First stab at "new" middleware handling.

### PR DESCRIPTION
**Pull request by apollo13**
https://github.com/django/deps/blob/master/draft/0005-rethinking-middleware.rst

### CODE REVIEW

File: django/core/handlers/base.py [hunk 1]
```diff
@@ -69,6 +75,11 @@ def load_middleware(self):
                 self._response_middleware.insert(0, mw_instance.process_response)
             if hasattr(mw_instance, 'process_exception'):
                 self._exception_middleware.insert(0, mw_instance.process_exception)
+            if callable(mw_instance):
+                self._middleware.insert(0, mw_instance)
+
+        if middleware_count == len(self._middleware):
```
Comment (carljm) [inline django/core/handlers/base.py]: I don't understand how this check implies that all middleware are new-style. It seems that it would be true anytime no middleware raised `MiddlewareNotUsed`, which seems probably pretty common even with old-style middleware. What am I missing?

File: django/core/handlers/middleware.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class MiddlewareMixin(object):
+    def __call__(self, request, response_factory):
+        response = None
+        if hasattr(self, 'process_request'):
+            response = self.process_request(request)
+        if not response:
+            response = response_factory(request)
```
Comment (carljm) [inline django/core/handlers/middleware.py]: For completeness, shouldn't there also be a try/except here and a call to `process_exception` if an exception occurs?

**Comment by carljm** [comment]
So I guess this PR converts only `SessionMiddleware` as a POC, but if we were going to move forward with this we'd probably want to convert all the built-in middleware?

**Comment by carljm** [comment]
I think this style of "function wrapper" middleware is simpler, easier to write, and easier to reason about correctly than the current middleware (where you can never be sure which of your middleware methods will actually be called, and thus e.g. we can't implement a reliable `TransactionMiddleware`). So in general I am +1 on this change. (I think adding a few draft docs to this PR might help make the nature and benefits of functional middleware a lot clearer to those who weren't part of our DUTH sprint discussions :-) ).

The part I'm not quite happy with yet is the transition plan. Currently in this PR, middlewares are all treated as "old-style" unless they are all "new-style" (though I don't understand how the detection of that in this PR is supposed to work). This is because we can't intersperse old-style and new-style middleware in a single ordered list and get the correct semantics for both. I think this upgrade mechanism is too subtle and error-prone (no way to easily tell by looking at your `MIDDLEWARE_CLASSES` setting whether all those middleware are new-style or not -- you'd have to go manually inspect each one).

I'd also really like to see the ability for new-style middleware to simply be a function, not a class with a `__call__` method that is implicitly instantiated at startup time. This means that the setting name `MIDDLEWARE_CLASSES` would become misleading. Also, I don't see any good way to autodetect the difference between the two in a single setting list, since people could have existing old-style middleware implemented as a factory function that they expect to be called at startup time.

For these reasons, I think a better and more explicit upgrade plan would be to introduce a new setting, `MIDDLEWARE`, and deprecate (maybe on a very slow deprecation path) `MIDDLEWARE_CLASSES`. If `MIDDLEWARE` is set, it should be a list of exclusively new-style middleware, and `MIDDLEWARE_CLASSES` is ignored. Otherwise `MIDDLEWARE_CLASSES` is handled the same as today.

The next big question is how to handle view middleware and template-response middleware. The other three kinds of middleware (request, response, and exception) can all be easily replaced by new-style middleware (request middleware is what you do in the top half of your new-style function, before calling up the chain to get the response; exception middleware can be implemented by wrapping that call in try/except, and response middleware is whatever you do in the bottom half, after getting the response.) But template-response and view middleware also have their uses, and aren't inherently replaced by new-style middleware.

I think the simplest and clearest option here is to also introduce a `VIEW_MIDDLEWARE` setting, which is a list of simple functions that behave the same as `process_view` does today, and `TEMPLATE_RESPONSE_MIDDLEWARE`, a list of simple functions that behave the same as `process_template_response` does today. I think that in practice it's rare to find a single middleware that implements both `process_view` or `process_template_response` and other methods, so I think in most cases this separation in settings actually makes it clearer what hooks are operating at various points in the cycle than it is today, where you have no idea (without deeper inspection) where a given piece of middleware might be hooking in.

I also think that really these latter two aren't middleware at all (they don't wrap anything, they just hook in at a certain point in the request/response cycle), so perhaps something like `VIEW_HOOK` and `TEMPLATE_RESPONSE_HOOK` would be better names?

The biggest downside to all this, of course, is that it's another big breaking change (not immediately breaking, but requiring update at some point), and there's been some sense that maybe we've had too many of those recently. Something like `MiddlewareMixin` could ease the upgrade somewhat. Another option is not to deprecate old-style `MIDDLEWARE_CLASSES` at all, and just let the two coexist indefinitely.

</brain-dump>

File: django/core/handlers/base.py (hunk 1 above)
Comment (apollo13) [inline django/core/handlers/base.py]: You are missing that `self._middleware` is only set if the middleware is callable, ergo a newstyle one…

File: django/core/handlers/middleware.py (hunk 2 above)
Comment (apollo13) [inline django/core/handlers/middleware.py]: No, since the existing code performs like that -- for better or worse (`response_factory` will call it internally with the same semantics as currently, unless I mixd that up).

**Comment by apollo13** [comment]
> So I guess this PR converts only SessionMiddleware as a POC, but if we were going to move forward with this we'd probably want to convert all the built-in middleware?

Yes, that was just to see into how many issues I'd run :D

**Comment by apollo13** [comment]
Regarding your novel; I don't really care -- adding three new settings strikes me as somewhat ugly, but I agree that the resulting functions would be nicer than those abnormities of middleware classes. That said you seem to be missing one minor important point of the classes: their call to the constructor allows them to perform "performance intensitive" tasks or to disable them during startup. How would that work with functions?

That said, I'll happily implement what you/(technical board) decides, in the end I think we just need a decision on how to continue…

File: django/core/handlers/base.py (hunk 1 above)
Comment (carljm) [inline django/core/handlers/base.py]: Ha yes! I see that now.

**Comment by carljm** [comment]
Yeah, good point about where to put initialization code. I think this is probably something that should be discussed on -developers. I'll think it through a bit more and then post something there about the options. I'll try to keep the novels on the mailing list instead of on GitHub :-)

File: django/core/handlers/middleware.py (hunk 2 above)
Comment (carljm) [inline django/core/handlers/middleware.py]: Yes, you're right, I see that for purposes of backwards-compatibility with old-style middleware (which is the point of this mix-in), this is adequate as-is.

**Comment by apollo13** [comment]
@carljm Did that email got lost in my inbox or did you never get around it?

**Comment by carljm** [comment]
The latter. I'll write it today.

**Comment by carljm** [comment]
Actually, strike that. I don't think two days before Christmas is probably the best time, on average, to get people's attention on a complex issue on -developers. I'll do it after the new year.

**Comment by carljm** [comment]
I'm working on writing up some API options here in the form of a DEP.

**Comment by apollo13** [comment]
Closing in favour of #5949

File: django/conf/global_settings.py [hunk 3]
```diff
@@ -427,6 +427,8 @@
     'django.middleware.csrf.CsrfViewMiddleware',
 ]
 
+MIDDLEWARES = None
```
Comment (davidszotten) [inline django/conf/global_settings.py]: isn't the plural of middleware still  middleware?

Comment (apollo13) [inline django/conf/global_settings.py]: Ha, no idea :D That might be the reason @carljm called it `MIDDLEWARE` in the first place ;)

Comment (carljm) [inline django/conf/global_settings.py]: Yeah, I addressed this in some detail here: https://groups.google.com/d/msg/django-developers/8LMJ44KAxWI/XtnRW2GwEgAJ

tl;dr I think either is defensible, as usage of the term "middleware" is fluid. I prefer `MIDDLEWARE` as the plural, and for now I left it that way in the spec; but I don't feel super-strongly about it.
